### ISSUE TITLE
Feature Request: numbered entities with custom counters

### ISSUE THREAD

**Issue by ararunaufc** [typst/typst#325]
Having a way to provide a counter for an `enum` to use would be great for one use case I have right now, which involves defining a custom numbered list, but it would also enable resuming the counting of one `enum` from another, like when you "suspend" the list with some text in between items.

Replicating this idea on other numbered entities would also easen customization, like defining separate numbering on front matter / main matter / back matter, or continuous numbering for Theorems, Lemmas, etc.

**Comment by lingo** [comment]
I have a different use case, but I believe this is relevant here.

I feel that we should be able to provide a counter for an enum, or be able to retrieve the value of a counter from an enum.

*My use case: * I'm trying to create a link to a particular step of an enum.

I can use a link successfully with a label, but I wondered if I can auto-generate the text from the enum's counter, e.g. See #link(<important-step>)[...].  I tried using a `ref` but this of course currently requires a figure. 

I tried to use locate() and query() to display a counter value, but when trying to use counter(enum),  I get `error: enum is not locatable`
I tried using a custom figure kind, and my own counter, but had problems there too.

 From looking at the typst source, it appears `enum`s indeed do not use counters. Perhaps they should, like headings? The logic seems to be sufficiently similar (e.g. use of numbering vs level, etc)

**Comment by laurmaedje** [comment]
Better integration between enums and counters seems very reasonable. One big difference though is that counters are always global to the document whereas enums are generally counted locally.

**Comment by AlbertoFabbri93** [comment]
I ended up here by searching how to skip a number when using a numbered list. The idea is that something like this:
```typst
#set enum(start: 2)

+ Bla

+ Bla bla

#counter(enum).step()

+ Bla bla bla
```
Should be printed as:
2. Bla
3. Bla bla
5. Bla bla bla

**Comment by Andrew15-5** [comment]
I had a very similar problem to the one of @lingo, but my workaround didn't work perfectly:

<details><summary>source</summary>
<p>

```typ
#locate(loc => counter("source").final(loc).first())
#show enum.item: it => {
  counter("source").step()
  it
}
+ item
+ item
+ item
```

</p>
</details>

<details><summary>output</summary>
<p>

![image](https://github.com/typst/typst/assets/37143421/30aaeeed-eb5e-40e6-b9ef-43f157f257bc)

</p>
</details>

Using `set enum(numbering)` instead of `show enum.item` works flawlessly:

<details><summary>source</summary>
<p>

```typ
#locate(loc => counter("source").final(loc).first())
#set enum(full: false, numbering: (n) => {
  counter("source").update(n)
  numbering("1.", n)
})
+ item
+ item
+ item
```

</p>
</details>

<details><summary>output</summary>
<p>

![image](https://github.com/typst/typst/assets/37143421/3f590c68-7d27-4d35-a0bc-fecf08fa0e24)

</p>
</details>

**Comment by sbaumard** [comment]
I can't seem to be able to adapt @Andrew15-5 's workaround to both manage multi-leveled numbered lists (such as "1)a)") and reuse the saved counter in a new list. Does anyone have an idea ?

**Comment by Andrew15-5** [comment]
Yeah, multi-level numbering is not as simple. I haven't tried it with that, but it might work with some improved version. For that, the argument must be `..n` instead of `n`, that is for sure.

**Comment by Esn024** [comment]
>` query(<step-link>, before: loc).first()`

This part of @lingo's example currently gives the error "missing argument: location". How should one write this with the current syntax?

I'm trying to figure out if the above examples from lingo and @Andrew15-5 will work with the solution given [here](https://gist.github.com/PgBiel/23a116de4a235ad4cf6c7a05d6648ca9) to allow enums to be referenced later in the document (that solution works, but only if you number your numbered list manually as in the example
```
1. amongus <label>
2. amongus
3. trimongus <labelled>
```
because the `number` property in typst only seems to be set if you number your list manually, not automatically.

...which makes that solution a lot less useful, and also seems to make it impossible to reference those items in a way not using numbers (roman numerals, letters, etc.)

**Comment by laurmaedje** [comment]
The 0.10 (currently released) syntax is `query(selector(<step-link>).before(loc), loc)`. With the upcoming 0.11, it will become `query(selector(<step-link>).before(here()))` and should be wrapped in a `context` block instead of a `locate` callback.

**Comment by Andrew15-5** [comment]
Will the current usage/syntax be deprecated or removed completely?

**Comment by Esn024** [comment]
@laurmaedje, thank you, that works. Is there a way to put @lingo's solution into a function, though, so that it's not so wordy for every reference to a label?

I tried 

```
let l (label, numbering: "1.") = {
  link(label)[#locate(loc => {
      let item = query(selector(label).before(loc), loc).first();
      [#{numbering(numbering,..s.at(item.location()))}]
  })]
}
```
But it says "only element functions can be used as selectors".

If I can get syntax of the function correct, I'd test it like so:
```
#[
  // create a counter
  #let s = counter("number")
  #{ s.update(c => 1) } // start at 1

  // ensure all enums (in this block?) update the counter before rendering
  #set enum(numbering: it  => {
    s.update(number => it)
    numbering("1.", it) // can't seem to find out what numbering the enum is using
  })

  Text
  
  + point 1
  + point 2 <step-link2>

Reference above point: #l(<step-link2>, "A)")
]
```

Ultimately it would be nice to use `@step-link2`, but I'm just looking for something that works and is relatively concise for now.

**Comment by IsoNovio** [comment]
I figured out two very simple ways to create an enum that continues enumerating across the entire document, similar to what LaTeX's gb4e does.

The first way is to use a function.
```typst
#let ex(..items) = enum(
  numbering: (n) => context {
    let ex-counter = counter("ex")
    ex-counter.step()
    let ex-count = ex-counter.get().at(0) + 1
    numbering("(1)", ex-count)
  },
  ..items
)
/* usage:
#ex[example item][another item]
*/
```

The set way is to use `set` on `enum`
```typst
#set enum(
  full: true,
  numbering: (..nums-arguments) => context {
    let nums = nums-arguments.pos()
    if nums.len() == 1 {
      let thread-counter = counter("ex")
      thread-counter.step()
      let thread-count = thread-counter.get().first() + 1
      numbering("(1)", thread-count)
    } else {
      numbering("1.", nums.last())
    }
  },
)
/* usage:
+ example item
*/
```

**Comment by tianyi-smile** [comment]
Now, the package `itemize:0.2.0`  supports resuming enum numbering. To enable this feature, set the `auto-resuming` parameter to `auto` in the method `default-enum-list` (or `default-enum`). 

- Use the `resume()` method to continue numbering from the previous ones.

  ```typst
  #import "@preview/itemize:0.2.0" as el
  #el.default-enum-list(auto-resuming: auto)[
    + #lorem(5)
      + #lorem(5)
      + #lorem(5)
      #lorem(5)
    + #lorem(5)
      #el.resume() // -> 3
      + #lorem(5)
      + #lorem(5)
    + #lorem(5)
      + + #lorem(5)
  ]
  ```

  
<img width="1191" height="397" alt="Image" src="https://github.com/user-attachments/assets/5c46e625-16bf-4af4-a632-4cccf67a905b" />

- Alternatively, use `resume-label(<some-label>)` to label the enum to resume, then use `resume-list(<some-label>)` in the desired enum to continue numbering.

  - If the following is added to the document:

    ```typst
    #show: el.config.ref-resume
    ```

    we can use `@some-label` instead of `resume-list(<some-label>)`.

  ```typst
  #import "@preview/itemize:0.2.0" as el
  #show: el.config.ref-resume
  #let auto-resume = el.default-enum-list.with(auto-resuming: auto)
  #auto-resume[
    + #lorem(5)
      + #lorem(5)
      + #lorem(5)
      - #lorem(5)
      #el.resume() // continue
      + #lorem(5) #el.resume-label(<resume:demo>)
    #lorem(5)
    @resume:demo // resume the enum labelled with `resume:demo`
    + #lorem(5)
  ]
  ```

  
<img width="1191" height="363" alt="Image" src="https://github.com/user-attachments/assets/f22bc4ab-6b17-41b2-bc38-6deb8ba3d0b5" />

- Alternatively, use `auto-resume-enum(auto-resuming: true)[...]` to ensure all `enum` items within `[...]` continue numbering from the previous items. For example:

  ```typst
  #import "@preview/itemize:0.2.0" as el
  #let resume-enum(doc) = el.default-enum-list(auto-resuming: auto)[
    #el.auto-resume-enum(auto-resuming: true, doc)
  ]
  #resume-enum[
    + #lorem(5)
        + #lorem(5)
        + #lorem(5)
          + #lorem(5)
        - #lorem(5)
        + #lorem(5)
          + #lorem(5)
    #lorem(5)
    + #lorem(5)
  ]
  ```

  
<img width="1191" height="421" alt="Image" src="https://github.com/user-attachments/assets/0693285c-b880-4e31-bde9-ef9ed3a85367" />
